src/shared/json-queue.c

```diff
@@ -105,8 +133,260 @@ cJSON * jqueue_next(file_queue * queue) {
+            fclose(fileq->fp);
+            fileq->fp = NULL;
+            return -1;
+        }
+    }
+
+    /* File change time */
+    if (fileq->fp) {
+        if (fstat(fileno(fileq->fp), &fileq->f_status) < 0) {
+            merror(FSTAT_ERROR, fileq->file_name, errno, strerror(errno));
+            fclose(fileq->fp);
+            fileq->fp = NULL;
+            return -1;
+        }
+    }
+
+    fileq->last_change = fileq->f_status.st_mtime;
+
+    return 1;
+}
+
 // Close queue
 void jqueue_close(file_queue * queue) {
     fclose(queue->fp);
     queue->fp = NULL;
 }
+
+/* Set queue flags */
+void jqueue_flags(file_queue *fileq, int flags) {
+    fileq->flags = flags;
+}
+
+/* Return alert data for the next file in the queue */
+alert_data *GetAlertJSONData(file_queue *fileq) {
+    alert_data *al_data;
+    cJSON *al_json;
+    char *groups;
+    int i = 0;
+
+    cJSON *json_object;
+    cJSON *timestamp;
+    cJSON *rule;
+    cJSON *syscheck;
+    cJSON *location;
+    cJSON *srcip;
+    cJSON *full_log;
+
+    os_calloc(1, sizeof(alert_data), al_data);
+
+    /* Get message if available */
+    al_json = jqueue_next(fileq);
+
+    if (!al_json || !fileq->fp) {
+        cJSON_Delete(al_json);
+        FreeAlertData(al_data);
+        return NULL;
+    }
+
+    /* Date */
+    timestamp = cJSON_GetObjectItem(al_json, "timestamp");
+
+    if (timestamp) {
+        os_strdup(timestamp->valuestring, al_data->date);
+    }
+
+    /* Rule */
+    rule = cJSON_GetObjectItem(al_json, "rule");
+
+    if (!rule) {
+        cJSON_Delete(al_json);
+        FreeAlertData(al_data);
+        return NULL;
+    }
+
+    // Rule ID
+    json_object = cJSON_GetObjectItem(rule, "id");
+
+    if (json_object) {
+        al_data->rule = atoi(json_object->valuestring);
+    }
+
+    // Rule description
+    json_object = cJSON_GetObjectItem(rule, "description");
+
+    if (json_object) {
+        os_strdup(json_object->valuestring, al_data->comment);
+    }
+
+    // Groups
+    json_object = cJSON_GetObjectItem(rule, "groups");
+
+    if (json_object) {
+        /* Groups is an array in the alerts.json file */
+        /*
+            First, we copy the first item in groups, then the rest,
+            in case there is more than one
+        */
+        os_calloc(1, strlen(cJSON_GetArrayItem(json_object, 0)->valuestring) + 1, groups);
+        strcpy(groups, cJSON_GetArrayItem(json_object, 0)->valuestring);
+
+        for (i = 1; i < cJSON_GetArraySize(json_object); i++) {
+            os_realloc(groups, strlen(groups) + strlen(cJSON_GetArrayItem(json_object, i)->valuestring) + 2, groups);
+            strcat(groups, ",");
+            strcat(groups, cJSON_GetArrayItem(json_object, i)->valuestring);
+        }
+
+        os_strdup(groups, al_data->group);
+
+        os_free(groups);
+    }
+
+    // Level
+    json_object = cJSON_GetObjectItem(rule, "level");
+
+    if (json_object) {
```

Comment:
Same here, error/warning if a rule doesn't contain the level.

Reply:
Done: https://github.com/wazuh/wazuh/pull/4256/files#diff-4389a0705fc67fca0285ad7bf1614e46R289